File: src/gui/objects/grid/GridView.java
/**
 * Copyright (c) 2013, 2015, ControlsFX
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *     * Neither the name of ControlsFX, any associated website, nor the
 * names of its contributors may be used to endorse or promote products
 * derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL CONTROLSFX BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package gui.objects.grid;


import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.function.Predicate;

import javafx.beans.property.DoubleProperty;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.collections.ObservableList;
import javafx.collections.transformation.FilteredList;
import javafx.collections.transformation.SortedList;
import javafx.css.*;
import javafx.event.Event;
import javafx.scene.Node;
import javafx.scene.control.Cell;
import javafx.scene.control.Control;
import javafx.scene.control.ListCell;
import javafx.scene.input.ScrollEvent;
import javafx.util.Callback;

import util.access.V;
import util.functional.Functors.Ƒ1;

import static java.util.Collections.unmodifiableList;
import static java.util.stream.Collectors.toCollection;
import static javafx.collections.FXCollections.observableArrayList;
import static util.async.Async.runLater;
import static util.functional.Util.stream;

/**
 * A GridView is a virtualised control for displaying items in a
 * visual, scrollable, grid-like fashion. In other words, whereas a ListView
 * shows one {@link ListCell} per row, in a GridView there will be zero or more
 * {@link GridCell} instances on a single row.
 * <p/>
 * This approach means that the number of GridCell instances
 * instantiated will be a significantly smaller number than the number of
 * items in the GridView items list, as only enough GridCells are created for
 * the visible area of the GridView. This helps to improve performance and
 * reduce memory consumption.
 * <p/>
 * Because each {@link GridCell} extends from {@link Cell}, the same approach
 * of cell factories that is taken in other UI controls is also taken in GridView.
 * This has two main benefits:
 *
 * <ol>
 *   <li>GridCells are created on demand and without user involvement,
 *   <li>GridCells can be arbitrarily complex. A simple GridCell may just have
 *   its {@link GridCell#textProperty() text property} set, whereas a more complex
 *   GridCell can have an arbitrarily complex scenegraph set inside its
 *   {@link GridCell#graphicProperty() graphic property} (as it accepts any Node).
 * </ol>
 *
 *
 *
 * @see GridCell
 */
public class GridView<T,F> extends Control {

    final Class<F> type;
    final ObservableList<T> itemsAll;
    final FilteredList<T> itemsFiltered;
    final SortedList<T> itemsSorted;
    final Ƒ1<T,F> filterByMapper;


    /*
     * Predicate that filters the table list. Null predicate will match all
     * items (same as always true predicate). The value reflects the filter
     * generated by the user through the {@link #searchBox}. Changing the
     * predicate programmatically is possible, however the searchBox will not
     * react on the change, its effect will merely be overridden and when
     * search box predicate changes, it will in turn override effect of a
     * custom predicate.
     */
    public final ObjectProperty<Predicate<? super T>> itemsPredicate;
    public final ObjectProperty<Comparator<? super T>> itemsComparator;

    public final V<T> selectedItem = new V<>(null);
    public final V<T> selectedRow = new V<>(null);

    private boolean scrollFlag = true;

    /** Creates a default, empty GridView control. */
    public GridView(Class<F> type) {
        this(type, null, null);
    }

    /** Creates a default, empty GridView control. */
    public GridView(Class<F> type, Ƒ1<T,F> filterByMapper) {
        this(type, filterByMapper, null);
    }

    /** Convenience consturctor. Creates an empty GridView with specified sizes. */
    public GridView(Class<F> type, Ƒ1<T,F> filterByMapper, double cellWidth, double cellHeight, double vgap, double hgap) {
        this(type, filterByMapper, null);
        setCellWidth(cellWidth);
        setCellHeight(cellHeight);
        setHorizontalCellSpacing(hgap);
        setVerticalCellSpacing(vgap);
    }

    /** Convenience constructor. Creates a default GridView with the provided items. */
    public GridView(Class<F> type, Ƒ1<T,F> filterByMapper, ObservableList<T> backingList) {
        this.type = type;
        this.filterByMapper = filterByMapper==null ? x -> (F)x : filterByMapper;

        itemsAll = backingList==null ? observableArrayList() : backingList;
        itemsFiltered = new FilteredList<>(itemsAll);
        itemsSorted = new SortedList<>(itemsFiltered);

        itemsPredicate = itemsFiltered.predicateProperty();
        itemsComparator = itemsSorted.comparatorProperty();

        getStyleClass().add(DEFAULT_STYLE_CLASS);

        // Decrease scrolling speed
        // The default scrolling speed is simply too much. On my system its more than a full
        // vertical 'view' which is very confusing as user loses any indication of scrolling amount.
        // impl: consume scroll events and refire with smaller vertical values
        double factor = 1/3d;
        addEventFilter(ScrollEvent.ANY, e -> {
            if(scrollFlag) {
                Event ne = new ScrollEvent(
                    e.getEventType(),e.getX(),e.getY(),e.getScreenX(),e.getScreenY(),
                    e.isShiftDown(),e.isControlDown(),e.isAltDown(),e.isMetaDown(),e.isDirect(),
                    e.isInertia(),e.getDeltaX(),e.getDeltaY()*factor,e.getTextDeltaX(),e.getTextDeltaY()*factor,
                    e.getTextDeltaXUnits(),e.getTextDeltaX(),e.getTextDeltaYUnits(),e.getTextDeltaY()*factor,
                    e.getTouchCount(),e.getPickResult()
                );
                e.consume();
                scrollFlag = false;
                runLater(() -> {
                    if (e.getTarget() instanceof Node) {
                        ((Node) e.getTarget()).fireEvent(ne);
                    }
                    scrollFlag = true;
                });
            }
        });
    }

    @Override
    protected GridViewSkin<T,F> createDefaultSkin() {
        return new GridViewSkin<>(this);
    }

    @SuppressWarnings("unchecked")
    public GridViewSkin<T,F> getSkinn() {
        return (GridViewSkin) getSkin();
    }


    public ObservableList<T> getItemsRaw() {
        return itemsAll;
    }

    public ObservableList<T> getItemsShown() {
        return itemsSorted;
    }



    /**
     * Property for specifying how much spacing there is between each cell
     * in a row (i.e. how much horizontal spacing there is).
     */
    public final DoubleProperty horizontalCellSpacingProperty() {
        if (horizontalCellSpacing == null) {
            horizontalCellSpacing = new StyleableDoubleProperty(12) {
                @Override public CssMetaData<GridView<?,?>, Number> getCssMetaData() {
                    return GridView.StyleableProperties.HORIZONTAL_CELL_SPACING;
                }

                @Override public Object getBean() {
                    return GridView.this;
                }

                @Override public String getName() {
                    return "horizontalCellSpacing";
                }
            };
        }
        return horizontalCellSpacing;
    }
    private DoubleProperty horizontalCellSpacing;

    /**
     * Sets the amount of horizontal spacing there should be between cells in
     * the same row.
     * @param value The amount of spacing to use.
     */
    public final void setHorizontalCellSpacing(double value) {
        horizontalCellSpacingProperty().set(value);
    }

    /**
     * Returns the amount of horizontal spacing there is between cells in
     * the same row.
     */
    public final double getHorizontalCellSpacing() {
        return horizontalCellSpacing == null ? 12.0 : horizontalCellSpacing.get();
    }



    /**
     * Property for specifying how much spacing there is between each cell
     * in a column (i.e. how much vertical spacing there is).
     */
    private DoubleProperty verticalCellSpacing;
    public final DoubleProperty verticalCellSpacingProperty() {
        if (verticalCellSpacing == null) {
            verticalCellSpacing = new StyleableDoubleProperty(12) {
                @Override public CssMetaData<GridView<?,?>, Number> getCssMetaData() {
                    return GridView.StyleableProperties.VERTICAL_CELL_SPACING;
                }

                @Override public Object getBean() {
                    return GridView.this;
                }

                @Override public String getName() {
                    return "verticalCellSpacing";
                }
            };
        }
        return verticalCellSpacing;
    }

    /**
     * Sets the amount of vertical spacing there should be between cells in
     * the same column.
     * @param value The amount of spacing to use.
     */
    public final void setVerticalCellSpacing(double value) {
        verticalCellSpacingProperty().set(value);
    }

    /**
     * Returns the amount of vertical spacing there is between cells in
     * the same column.
     */
    public final double getVerticalCellSpacing() {
        return verticalCellSpacing == null ? 12.0 : verticalCellSpacing.get();
    }



    /**
     * Property representing the width that all cells should be.
     */
    public final DoubleProperty cellWidthProperty() {
        if (cellWidth == null) {
            cellWidth = new StyleableDoubleProperty(64) {
                @Override public CssMetaData<GridView<?,?>, Number> getCssMetaData() {
                    return GridView.StyleableProperties.CELL_WIDTH;
                }

                @Override public Object getBean() {
                    return GridView.this;
                }

                @Override public String getName() {
                    return "cellWidth";
                }
            };
        }
        return cellWidth;
    }
    private DoubleProperty cellWidth;

    /**
     * Sets the width that all cells should be.
     */
    public final void setCellWidth(double value) {
        cellWidthProperty().set(value);
    }

    /**
     * Returns the width that all cells should be.
     */
    public final double getCellWidth() {
        return cellWidth == null ? 64.0 : cellWidth.get();
    }


    /**
     * Property representing the height that all cells should be.
     */
    public final DoubleProperty cellHeightProperty() {
        if (cellHeight == null) {
            cellHeight = new StyleableDoubleProperty(64) {
                @Override public CssMetaData<GridView<?,?>, Number> getCssMetaData() {
                    return GridView.StyleableProperties.CELL_HEIGHT;
                }

                @Override public Object getBean() {
                    return GridView.this;
                }

                @Override public String getName() {
                    return "cellHeight";
                }
            };
        }
        return cellHeight;
    }
    private DoubleProperty cellHeight;

    /**
     * Sets the height that all cells should be.
     */
    public final void setCellHeight(double value) {
        cellHeightProperty().set(value);
    }

    /**
     * Returns the height that all cells should be.
     */
    public final double getCellHeight() {
        return cellHeight == null ? 64.0 : cellHeight.get();
    }


    private ObjectProperty<Callback<GridView<T,F>, GridCell<T,F>>> cellFactory;

    /**
     * Property representing the cell factory that is currently set in this
     * GridView, or null if no cell factory has been set (in which case the
     * default cell factory provided by the GridView skin will be used). The cell
     * factory is used for instantiating enough GridCell instances for the
     * visible area of the GridView. Refer to the GridView class documentation
     * for more information and examples.
     */
    public final ObjectProperty<Callback<GridView<T,F>, GridCell<T,F>>> cellFactoryProperty() {
        if (cellFactory == null) {
            cellFactory = new SimpleObjectProperty<>(this, "cellFactory");
        }
        return cellFactory;
    }

    /**
     * Sets the cell factory to use to create {@link GridCell} instances to
     * show in the GridView.
     */
    public final void setCellFactory(Callback<GridView<T,F>, GridCell<T,F>> value) {
        cellFactoryProperty().set(value);
    }

    /**
     * Returns the cell factory that will be used to create {@link GridCell}
     * instances to show in the GridView.
     */
    public final Callback<GridView<T,F>, GridCell<T,F>> getCellFactory() {
        return cellFactory == null ? null : cellFactory.get();
    }



    private static final String DEFAULT_STYLE_CLASS = "grid-view";

    private static class StyleableProperties {
        private static final CssMetaData<GridView<?,?>,Number> HORIZONTAL_CELL_SPACING =
            new CssMetaData<>("-fx-horizontal-cell-spacing", StyleConverter.getSizeConverter(), 12d) {

            @Override public Double getInitialValue(GridView<?,?> node) {
                return node.getHorizontalCellSpacing();
            }

            @Override public boolean isSettable(GridView<?,?> n) {
                return n.horizontalCellSpacing == null || !n.horizontalCellSpacing.isBound();
            }

            @Override
            @SuppressWarnings("unchecked")
            public StyleableProperty<Number> getStyleableProperty(GridView<?,?> n) {
                return (StyleableProperty<Number>)n.horizontalCellSpacingProperty();
            }
        };

        private static final CssMetaData<GridView<?,?>,Number> VERTICAL_CELL_SPACING =
            new CssMetaData<>("-fx-vertical-cell-spacing", StyleConverter.getSizeConverter(), 12d) {

            @Override public Double getInitialValue(GridView<?,?> node) {
                return node.getVerticalCellSpacing();
            }

            @Override public boolean isSettable(GridView<?,?> n) {
                return n.verticalCellSpacing == null || !n.verticalCellSpacing.isBound();
            }

            @Override
            @SuppressWarnings("unchecked")
            public StyleableProperty<Number> getStyleableProperty(GridView<?,?> n) {
                return (StyleableProperty<Number>)n.verticalCellSpacingProperty();
            }
        };

        private static final CssMetaData<GridView<?,?>,Number> CELL_WIDTH =
            new CssMetaData<>("-fx-cell-width", StyleConverter.getSizeConverter(), 64d) {

            @Override public Double getInitialValue(GridView<?,?> node) {
                return node.getCellWidth();
            }

            @Override public boolean isSettable(GridView<?,?> n) {
                return n.cellWidth == null || !n.cellWidth.isBound();
            }

            @Override
            @SuppressWarnings("unchecked")
            public StyleableProperty<Number> getStyleableProperty(GridView<?,?> n) {
                return (StyleableProperty<Number>)n.cellWidthProperty();
            }
        };

        private static final CssMetaData<GridView<?,?>,Number> CELL_HEIGHT =
            new CssMetaData<>("-fx-cell-height", StyleConverter.getSizeConverter(), 64d) {

            @Override public Double getInitialValue(GridView<?,?> node) {
                return node.getCellHeight();
            }

            @Override public boolean isSettable(GridView<?,?> n) {
                return n.cellHeight == null || !n.cellHeight.isBound();
            }

            @Override
            @SuppressWarnings("unchecked")
            public StyleableProperty<Number> getStyleableProperty(GridView<?,?> n) {
                return (StyleableProperty<Number>)n.cellHeightProperty();
            }
        };

        static final List<CssMetaData<? extends Styleable,?>> STYLEABLES = stream(Control.getClassCssMetaData())
                .append(HORIZONTAL_CELL_SPACING, VERTICAL_CELL_SPACING, CELL_WIDTH, CELL_HEIGHT)
                .collect(toCollection(() -> unmodifiableList(new ArrayList<>())));
    }

    /**
     * @return The CssMetaData associated with this class, which may include the
     * CssMetaData of its super classes.
     */
    public static List<CssMetaData<? extends Styleable, ?>> getClassCssMetaData() {
        return StyleableProperties.STYLEABLES;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public List<CssMetaData<? extends Styleable, ?>> getControlCssMetaData() {
        return getClassCssMetaData();
    }
}

File: src/player/main/sp/it/pl/ui/objects/grid/GridView.kt
package sp.it.pl.ui.objects.grid

import java.util.function.Predicate
import javafx.beans.property.ObjectProperty
import javafx.collections.FXCollections
import javafx.collections.FXCollections.observableSet
import javafx.collections.ObservableList
import javafx.collections.transformation.FilteredList
import javafx.collections.transformation.SortedList
import javafx.css.StyleConverter
import javafx.css.StyleableObjectProperty
import javafx.event.Event.ANY
import javafx.event.EventHandler
import javafx.scene.control.Control
import kotlin.reflect.KClass
import kotlin.streams.asSequence
import sp.it.pl.ui.objects.search.SearchAutoCancelable
import sp.it.util.access.StyleableCompanion
import sp.it.util.access.V
import sp.it.util.access.fieldvalue.ObjectField
import sp.it.util.access.fieldvalue.StringGetter
import sp.it.util.access.sv
import sp.it.util.access.svMetaData
import sp.it.util.functional.Functors.F1
import sp.it.util.functional.asIs
import sp.it.util.math.P
import sp.it.util.reactive.onChange
import sp.it.util.type.SealedEnum

/**
 * Two-dimensional virtualized control for displaying items in a two-dimensional grid.
 * It offers similar set of features as [sp.it.pl.ui.objects.table.FilteredTable], namely sorting, searching and filtering.
 *
 * This container is virtualized - the number of [sp.it.pl.ui.objects.grid.GridCell] instances depends on the number of
 * items displayed, not number in total. This helps to improve performance and reduce memory consumption.
 *
 * @param <T> type of item displayed in a cells of this grid
 * @param <F> type of object derived from item to filter items by
 * @see GridCell
 */
class GridView<T: Any, F: Any>(type: KClass<F>, filterMapper: F1<T, F>, backingList: ObservableList<T>): Control() {
   /** Type of object derived from item to filter items by. Used primarily for filter function provider. */
   val type: KClass<F> = type
   /** Initial filter criteria for the filter, used when filter is opened or additional filter added  */
   val filterMapper: F1<T, F> = filterMapper
   /** Initial filter criteria for the filter, used when filter is opened or additional filter added  */
   var filterPrimaryField: ObjectField<F, *>? = null
   /** Filtering for the items. Maintains [itemsPredicate]. */
   val search: Search = Search()
   /** Visibility of the bottom panel that controls information about items and menu bar. */
   val footerVisible = V(true)

   /** All items, without [itemsPredicate] and [itemsComparator] applied. See [itemsShown]. */
   val itemsRaw: ObservableList<T> = backingList
   /** Filtered items. Wraps [itemsRaw]. */
   val itemsFiltered: FilteredList<T> = FilteredList(itemsRaw)
   /**
    * [FilteredList.predicateProperty] of the [itemsFiltered].
    *
    * Null predicate will match all items (same as always true predicate).
    * The value reflects the filter generated by the user through the [search].
    *
    * Changing the predicate programmatically is possible, however the searchBox will not react on the change,
    * its effect will merely be overridden and when search box predicate changes, it will in turn override effect of a
    * custom predicate.
    */
   val itemsPredicate: ObjectProperty<Predicate<in T>?> = itemsFiltered.predicateProperty()
   /** Sorted items. Wraps [itemsFiltered]. */
   val itemsSorted: SortedList<T> = SortedList(itemsFiltered)
   /** [SortedList.comparatorProperty] of the [itemsSorted]. */
   val itemsComparator: ObjectProperty<Comparator<in T>> = itemsSorted.comparatorProperty()
   /** Visible items. [itemsRaw], with [itemsPredicate] and [itemsComparator] applied. Identical to [itemsSorted] */
   val itemsShown: ObservableList<T> = itemsSorted

   /**
    * Selection change activation strategies. Can be combined. Empty set will cause this grid to ignore selection.
    * By default, ([SelectionOn.MOUSE_CLICK],[SelectionOn.KEY_PRESS]).
    */
   val selectOn = observableSet(SelectionOn.MOUSE_CLICK, SelectionOn.KEY_PRESS)!!
   /** Selected item. */
   val selectedItem = V<T?>(null)
   /** Selected items. Intended for immediate consumption (it may be backed by an observable). */
   val selectedItems: Sequence<T>
      get() = sequenceOf(selectedItem.value).filterNotNull()
   /** [selectedItems] or [itemsShown] if none selected. Intended for immediate consumption (it may be backed by an observable). */
   val selectedOrAllItems: Sequence<T>
      get() = selectedItem.value?.let { sequenceOf(it) } ?: itemsShown.asSequence()

   /** [selectedOrAllItems] if orAll is true or [selectedItems]. Intended for immediate consumption (it may be backed by an observable). */
   fun getSelectedOrAllItems(orAll: Boolean): Sequence<T> = if (orAll) selectedOrAllItems else selectedItems

   /** Horizontal gap between grid cells */
   val horizontalCellSpacing: StyleableObjectProperty<Double> by sv(HORIZONTAL_CELL_SPACING)
   /** Horizontal gap between grid cells */
   val verticalCellSpacing: StyleableObjectProperty<Double> by sv(VERTICAL_CELL_SPACING)
   /** Cell width */
   val cellWidth: StyleableObjectProperty<Double> by sv(CELL_WIDTH)
   /** Cell height */
   val cellHeight: StyleableObjectProperty<Double> by sv(CELL_HEIGHT)
   /** Cell text height. Similar to [javafx.scene.control.TableView.fixedCellSize]. Has no effect by default but can be used to calculate [cellHeight]. */
   val cellHeightFixed: StyleableObjectProperty<Double> by sv(CELL_HEIGHT_FIXED)

   val cellWidthActual
      get() = cellWidth.value.takeIf { it!=CELL_SIZE_UNBOUND } ?: width

   /** Cell gap layout strategy. Default [CellGap.CENTER]. */
   val cellAlign = V<CellGap>(CellGap.CENTER)
   /** Maximum number of cells in a horizontal row. Null indicates no limit. Default null. */
   val cellMaxColumns = V<Int?>(null)
   /** Cell factory responsible for creating cells, Null indicates to skin to use its own default factory.  Default null. */
   val cellFactory = V<((GridView<T, F>) -> GridCell<T, F>)?>(null)

   init {
      isFocusTraversable = true
      styleClass += STYLE_CLASS

      // search
      search.installOn(this)
      addEventFilter(ANY) { if (search.isActive) search.updateSearchStyles() }
      itemsShown.onChange { if (search.isActive) search.updateSearchStyles() }

      onScroll = EventHandler { it.consume() }
   }

   val skinImpl: GridViewSkin<T, F>?
      get() = skin.asIs()

   val cellsShown: Sequence<GridCell<T, F>>
      get() = skinImpl?.cells?.asSequence().orEmpty()

   val cellsAll: Sequence<GridCell<T, F>>
      get() = skinImpl?.cellsAll?.asSequence().orEmpty()

   override fun createDefaultSkin(): GridViewSkin<T, F> = GridViewSkin(this)

   override fun getControlCssMetaData() = classCssMetaData

   /** Strategy for cell selection change activation behavior. */
   enum class SelectionOn {
      MOUSE_HOVER, MOUSE_CLICK, KEY_PRESS
   }

   /** Predefined cell sizes. */
   enum class CellSize(val width: Double) {
      SMALL_12(12.0),
      SMALL_24(24.0),
      SMALL_36(36.0),
      SMALL_48(48.0),
      SMALL_60(60.0),
      SMALL(80.0),
      SMALLER(120.0),
      NORMAL(160.0),
      LARGE(240.0),
      VERY_LARGE(400.0),
      GIANT(600.0),
      EXTREME(800.0)
   }

   /** Affects cell layout. */
   sealed interface CellGap {
      fun computeGap(grid: GridView<*, *>, width: Double, columns: Int): Double
      fun computeStartX(grid: GridView<*, *>, width: Double, columns: Int): Double
      companion object: SealedEnum<CellGap>(CellGap::class)

      /**
       * Will position cells at exact positions from the left. The gap will be constant, but the cells will not be
       * horizontally center aligned.
       */
      data object CENTER: CellGap {
         override fun computeGap(grid: GridView<*, *>, width: Double, columns: Int): Double = grid.verticalCellSpacing.value
         override fun computeStartX(grid: GridView<*, *>, width: Double, columns: Int): Double = RIGHT.computeStartX(grid, width, columns)/2.0
      }

      data object LEFT: CellGap {
         override fun computeGap(grid: GridView<*, *>, width: Double, columns: Int): Double = grid.verticalCellSpacing.value
         override fun computeStartX(grid: GridView<*, *>, width: Double, columns: Int): Double = 0.0
      }

      data object RIGHT: CellGap {
         override fun computeGap(grid: GridView<*, *>, width: Double, columns: Int): Double = grid.verticalCellSpacing.value
         override fun computeStartX(grid: GridView<*, *>, width: Double, columns: Int): Double = width - columns*grid.cellWidth.value - (columns - 1)*grid.verticalCellSpacing.value
      }

      /**
       * The cells will be horizontally center aligned, but the gap size will change depending on
       * the total row width and number of cells in a row.
       */
      data object JUSTIFY: CellGap {
         override fun computeGap(grid: GridView<*, *>, width: Double, columns: Int): Double = (width - columns*grid.cellWidth.value)/(columns - 1)
         override fun computeStartX(grid: GridView<*, *>, width: Double, columns: Int): Double = 0.0
      }
   }

   inner class Search: SearchAutoCancelable() {
      var field: StringGetter<F> = StringGetter.of<F?> { value, substitute -> value?.toString() ?: substitute }

      override fun doSearch(query: String) {
         for (i in itemsShown.indices) {
            val item = itemsShown[i]
            val itemS = if (item==null) null else field.getOfS(filterMapper.apply(item), "")
            if (itemS!=null && isMatchNth(itemS, query)) {
               skinImpl!!.select(i)
               updateSearchStyles()
               break
            }
         }
      }

      override fun cancel() {
         super.cancel()
         cellsAll.forEach {
            it.pseudoClassStateChanged(PC_SEARCH_MATCH, false)
            it.pseudoClassStateChanged(PC_SEARCH_MATCH_NOT, false)
         }
      }

      fun updateSearchStyles() {
         if (isCancelable.value) searchAutoCanceller.start(cancelActivityDelay.value)
         updateSearchStyleRowsNoReset()
      }

      // TODO: move to skin
      fun updateSearchStyleRowsNoReset() {
         val searchOn = isActive
         cellsShown.forEach { cell: GridCell<T, F> ->
            val item = cell.item
            val itemS = if (item==null) null else field.getOfS(filterMapper.apply(item), "")
            val isMatch = itemS!=null && isMatch(itemS, searchQuery.get())
            cell.pseudoClassStateChanged(PC_SEARCH_MATCH, searchOn && isMatch)
            cell.pseudoClassStateChanged(PC_SEARCH_MATCH_NOT, searchOn && !isMatch)
         }
      }
   }

   companion object: StyleableCompanion() {
      const val STYLE_CLASS = "grid-view"
      const val CELL_SIZE_UNBOUND = -1.0

      /** Creates an empty GridView with specified sizes. */
      inline operator fun <T: Any, reified F: Any> invoke(filterMapper: F1<T, F>, backingList: ObservableList<T>? = null): GridView<T, F> =
         GridView(F::class, filterMapper, backingList ?: FXCollections.observableArrayList())

      /** Creates an empty GridView with specified sizes. Convenience constructor. */
      inline operator fun <T: Any, reified F: Any> invoke(filterMapper: F1<T, F>, cellSize: P, gap: P, backingList: ObservableList<T>? = null): GridView<T, F> =
         GridView(filterMapper, backingList).apply {
            cellWidth.value = cellSize.x
            cellHeight.value = cellSize.y
            horizontalCellSpacing.value = gap.x
            verticalCellSpacing.value = gap.y
         }

      val HORIZONTAL_CELL_SPACING by svMetaData<GridView<*, *>, Double>("-fx-horizontal-cell-spacing", StyleConverter.getSizeConverter().asIs(), 12.0, GridView<*, *>::horizontalCellSpacing)
      val VERTICAL_CELL_SPACING by svMetaData<GridView<*, *>, Double>("-fx-vertical-cell-spacing", StyleConverter.getSizeConverter().asIs(), 12.0, GridView<*, *>::verticalCellSpacing)
      val CELL_WIDTH by svMetaData<GridView<*, *>, Double>("-fx-cell-width", StyleConverter.getSizeConverter().asIs(), 64.0, GridView<*, *>::cellWidth)
      val CELL_HEIGHT by svMetaData<GridView<*, *>, Double>("-fx-cell-height", StyleConverter.getSizeConverter().asIs(), 64.0, GridView<*, *>::cellHeight)
      val CELL_HEIGHT_FIXED by svMetaData<GridView<*, *>, Double>("-fx-fixed-cell-size", StyleConverter.getSizeConverter().asIs(), 12.0, GridView<*, *>::cellHeightFixed)
   }
}